appcypher-syntax.md

```diff
@@ -0,0 +1,947 @@
+## SOURCE FILE
+- Source file can be empty, otherwise it must contain a valid set of UTF-8 codepoints.
+
+- Support for UTF-8 identifiers is planned.
+
+
+## COMMENTS
+#### Single-line comments
+- Single-line comments start with `//`
+    ```js
+    // Hello, World!
+    ```
+
+#### Multiline comments
+- Nested multiline comment is allowed.
+    ```
+    /*
+     * Multiline comments can span multiple lines.
+     * They begin with `/*`.
+     * And they end with `*/`.
+     */
+    ```
+
+## SUBJECTS
+- Subject, in this context, refers to the identifier a value/object is bound to. We say a value/object is bound to a subject.
+
+    ```
+    let name = "Okechukwu"
+    ```
+
+    Here `name` is the subject and `"Okechukwu"` the value or object, so we say the value `"Okechukwu"` is bound to the subject `name`.
+
+- The `let` keyword is used for subject whose values cannot change.
+    ```js
+    let immutable = 10
+    immutable = 50 // Invalid!
+    ```
+
+- The `var` keyword is used for subject whose values can change.
+    ```js
+    var mutable = 10
+    mutable = 50 // Subject value changed to 50
+    ```
+
+- Late binding is not supported because it makes static analysis harder and there is no real benefit to it.
+    ```js
+    var mutable // Invalid!
+    ```
+
+#### Shadowing
+- A subject can be rebound to a different type within the same scope.
+    ```js
+    let phone = "08074789222" // `phone` has type `String` here
+    let phone = 08074789222 // `phone` now has type `Int`
+    ```
```

Comment:
Can you explain your thoughts here? Is there a reason shadowing constants doesn't make sense?

Reply:
I agree with the @Thecarisma . This kind of shadowing can be done for mutable variables only. Then the implementation of this will require the modification of the AST in a multi-pass sequence or something.

Reply:
No one has given any compelling reason though. 

I would like to know if shadowable constant can make compiler implementation hard or if it is going to lead to some unwanted consequences?

Constructive feedback is important.

Reply:
Shadowing constant will break the main concept of a constant itself, a constant should only be defined once. 

Also for the compiler implementation, we will have an extra rule of check on a variable that shadows a constant to ensure the scope is different instead of just breaking the process. 

 Also, it ok with me for an inner function to shadow a constant in the parameter but not in the body of the function.